File: go/0005_longest_palindromic_substring.go
package main

func longestPalindrome(s string) string {
	slen := len(s)
	var left, right int
	for i := 0; i < slen; i++ {
		l1, r1 := expandOdd(s, i)
		l2, r2 := expandEven(s, i)

		if r1-l1 >= right-left {
			left, right = l1, r1
		}
		if r2-l2 >= right-left {
			left, right = l2, r2
		}
	}
	return s[left : right+1]
}

func expandOdd(s string, i int) (left, right int) {
	for interval := 0; i-interval >= 0 && i+interval < len(s); interval++ {
		if s[i-interval] != s[i+interval] {
			break
		}
		left, right = i-interval, i+interval
	}
	return
}

func expandEven(s string, i int) (left, right int) {
	for interval := 0; i-interval >= 0 && i+interval+1 < len(s); interval++ {
		if s[i-interval] != s[i+interval+1] {
			break
		}
		left, right = i-interval, i+interval+1
	}
	return
}

File: python/0005_longest_palindromic_substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        maxSoFar = 0
        res = ''
        DP = [[0 for x in range(len(s))] for x in range(len(s))]
        for i in range(len(s)):
            DP[i][i] = True

            if maxSoFar < 1:
                res = s[i]
                maxSoFar = 1
            if i < len(s) - 1 and s[i] == s[i + 1]:
                DP[i][i + 1] = True
                if maxSoFar < 2:
                    res = s[i:i + 2]
                    maxSoFar = 2

        for width in range(3, len(s) + 1):
            for left in range(len(s)):
                right = left + width - 1
                if right >= len(s):
                    break
                if DP[left + 1][right - 1] and s[left] == s[right]:
                    DP[left][right] = True
                    if right - left + 1 > maxSoFar:
                        maxSoFar = right - left + 1
                        res = s[left:right + 1]
        return res